### ISSUE TITLE
Rendering: 

### ISSUE THREAD

**Issue by rahulshivajipawar** [libass/libass#691]
### Screenshots

<img width="736" alt="Screenshot 2023-05-02 at 8 11 05 PM" src="https://user-images.githubusercontent.com/7852108/235701402-196e8b37-5f5e-4e68-86ea-e7d95fda149f.png">

### Description of the issue

The subtitles in Kannada language for font type Noto Sans are not rendered correctly using ffmpeg's ass filter which in turn uses libass. 

Here is the link to get Noto Sans Kannada fonts. https://fonts.google.com/noto/specimen/Noto+Sans+Kannada 

### libass version

libass.so.9 

### ASS Sample

[Script Info]
; Script Generated by Banva
ScriptType: v4.00+
PlayResX: 384
PlayResY: 288
ScaledBorderAndShadow: yes

[V4+ Styles]
Format: Name, Fontname, Fontsize, PrimaryColour, SecondaryColour, OutlineColour, BackColour, Bold, Italic, Underline, StrikeOut, ScaleX, ScaleY, Spacing, Angle, BorderStyle, Outline, Shadow, Alignment, MarginL, MarginR, MarginV, Encoding
Style: Banva-1,Noto Sans,16,&H00ffffff,&Hffffff,&H00000000,&H000000,0,0,0,0,100,100,0,0,3,2,0,2,10,10,10,0

[Events]
Format: Layer, Start, End, Style, Name, MarginL, MarginR, MarginV, Effect, Text
Dialogue: 0,0:0:0.35,0:0:3.57,Banva-1,,0,0,0,,ಸ್ಮಿತ್ ಚಾರ್ಟ್ ಎಂದು ಕರೆಯಲ್ಪಡುವ ಮೈಕ್ರೋವೇವ್\Nಎಂಜಿನಿಯರಿಂಗ್ ನಲ್ಲಿ
Dialogue: 0,0:0:3.57,0:0:6.36,Banva-1,,0,0,0,,ನಾವು ಬಳಸುವ ಸಾಧನದ ಬಗ್ಗೆ ನಾನು ನಿಮಗೆ ಸ್ವಲ್ಪ\Nಹೇಳಲಿದ್ದೇನೆ.
Dialogue: 0,0:0:7.13,0:0:11.05,Banva-1,,0,0,0,,ಆದ್ದರಿಂದ ನೀವು ಅಡೆತಡೆಗಳ ಹೊಂದಾಣಿಕೆಯ ಬಗ್ಗೆ\Nಕಲಿತಿದ್ದೀರಿ.
Dialogue: 0,0:0:11.05,0:0:14.80,Banva-1,,0,0,0,,ಆದ್ದರಿಂದ ತಾಂತ್ರಿಕವಾಗಿ ಸಾಮಾನ್ಯವಾಗಿ\Nಮೈಕ್ರೊಸ್ಟ್ರಿಪ್
Dialogue: 0,0:0:14.80,0:0:18.22,Banva-1,,0,0,0,,ಲೈನ್ ಅಥವಾ ಯಾವುದೇ ರೀತಿಯ ರೇಖೆಯನ್ನು\Nಹೊಂದಿರುತ್ತದೆ.
Dialogue: 0,0:0:18.22,0:0:21.56,Banva-1,,0,0,0,,ಇಲ್ಲ, ಮತ್ತು ನೀವು ಆ ಎರಡನ್ನೂ\Nಹೊಂದಿಸುತ್ತಿದ್ದೀರಿ.
Dialogue: 0,0:0:22.69,0:0:23.03,Banva-1,,0,0,0,,ಉಹ್.
Dialogue: 0,0:0:23.81,0:0:28.22,Banva-1,,0,0,0,,ಲೋಡ್ ಆದ್ದರಿಂದ ನಿಮಗೆ ನೀಡುವ ಸೂತ್ರವನ್ನು ನಾವು\Nನೋಡಿದ್ದೇವೆ.
Dialogue: 0,0:0:28.94,0:0:29.97,Banva-1,,0,0,0,,ಇನ್ ಪುಟ್ ಇಂಪೆಡೆನ್ಸ್.root@ff336626527b:/src/245f0963-c3bc-4bdd-99f2-a6173a1ba7eb

### Special Fonts

I uploaded or linked to the required font

### System Information

ffmpeg burning of subtitles happens inside a linux (debian) docker container. 

System Info:

Linux ff336626527b 5.10.124-linuxkit #1 SMP PREEMPT Thu Jun 30 08:18:26 UTC 2022 aarch64 GNU/Linux

**Comment by TheOneric** [comment]
Thanks for your report! Unfortunately though, it is not immedeatily obvious to me what issue you are referring to and you will need to be a bit more precise.
Checking against Google’s online rendering nothing stands out to me on first glance:
![grafik](https://user-images.githubusercontent.com/1668471/235738774-d309c63e-7b2e-4807-b2c9-fb44d8b2c561.png)

 - Describe in detail what exactly appears wrong to you? What result were you expecting instead?
 - Replace `put-your-other-options-here` to match your example and then run `ffmpeg -v debug put-your-other-options-here 2>&1 | tee -a 691.log` and upload the resulting `691.log` file. This is important to see which fonts get actually selected and gives more info about your libass version.
 - Please provide the exact command line you were using
 - If you have access to a MS Windows installation, have you checked if VSFilter produces the same result? (ideally [xy-VSFilter](https://github.com/Cyberbeing/xy-VSFilter/releases/tag/3.0.0.306) and mention what version of Windows you were testing on)

**Comment by rahulshivajipawar** [comment]
The precise issue is that  background is not wrapped correctly around "spaces" between words and at the "period".

**Comment by rcombs** [comment]
So this is specific to BorderStyle 3, then?

**Comment by astiob** [comment]
Based on the screenshots, looks like a case where the named font doesn't contain Kannada (but does contain spaces and full stops) and a substitute font with different metrics is selected.We render BorderStyle 3 for each glyph, but it should be rendered for each glyph run using the main font's metrics. In fact, the metrics of glyphwise substitute fonts should not affect *anything*, not only BorderStyle 3. For example, line height/spacing suffers as well.

**Comment by rahulshivajipawar** [comment]
> So this is specific to BorderStyle 3, then?

Yes, this is specific to BorderStyle 3.

**Comment by rahulshivajipawar** [comment]
> 



> Based on the screenshots, looks like a case where the named font doesn't contain Kannada (but does contain spaces and full stops) and a substitute font with different metrics is selected.We render BorderStyle 3 for each glyph, but it should be rendered for each glyph run using the main font's metrics. In fact, the metrics of glyphwise substitute fonts should not affect *anything*, not only BorderStyle 3. For example, line height/spacing suffers as well.

You are correct. I tried changing the font type from "Noto Sans" to "Noto Sans Kannada", then all the glyphs are rendered correctly.  During the original rendering, "spaces" and "period" were probably rendered using "Noto Sans" font while the rest of the glyphs were rendered using the substitute font "Noto Sans Kannada."
<img width="720" alt="Screenshot 2023-05-03 at 9 05 28 AM" src="https://user-images.githubusercontent.com/7852108/235828184-839e0a5f-61bb-488b-8bd7-589ea12d7094.png">
